### PR TITLE
ExternalPetscSolverApp: implement backup and restore

### LINKED ISSUE #12595: An example using  PETSc as an external solver
## Rationale
MOOSE can couple an external app in a different way. A new way is to use `ExternalProblem` that
calls an external solver and transfers data back-forward.

It is necessary to have an example to demonstrate this capability

## Description
Create a PETSc example that uses an structured mesh to solve a simple time-dependent diffusion equation. We need to add a mirror of PETSc structured mesh DMDA in MOOSE.  And then we can copy solution back-forward between PETSc and MOOSE.

## Impact
It won't impact users.

### PR DESCRIPTION
**Pull request by fdkong**
Implemented backup and restore for the PETSc external solver. As a demo, the PETSc application needs to backup and restore only one solution. 

External App developers need to develop their own ways to backup and restore necessary data during Ricard iterations.

Closes #12595

### CODE REVIEW

File: modules/external_petsc_solver/include/problem/ExternalPETScProblem.h [hunk 1]
```diff
@@ -46,7 +46,6 @@ class ExternalPETScProblem : public ExternalProblem
   /// PETSc solver
   TS & _ts;
   /// PETSc solver solution
-  Vec _petsc_sol;
+  Vec & _petsc_sol;
```
Comment (lindsayad) [inline modules/external_petsc_solver/include/problem/ExternalPETScProblem.h]: Why this change? We need a reference to a pointer?

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C [hunk 2]
```diff
@@ -12,12 +12,15 @@
...
 
-template <>
 InputParameters
-validParams<ExternalPetscSolverApp>()
+ExternalPetscSolverApp::validParams()
 {
-  InputParameters params = validParams<MooseApp>();
+  InputParameters params = MooseApp::validParams();
+
+  // Do not use legacy DirichletBC, that is, set DirichletBC default for preset = true
```
Comment (lindsayad) [inline modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: This sentence is a little confusing. Maybe just say "By default, use preset BCs"

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C [hunk 3]
```diff
@@ -27,6 +30,13 @@ ExternalPetscSolverApp::ExternalPetscSolverApp(InputParameters parameters) : Moo
 
 #if LIBMESH_HAVE_PETSC
   PETScExternalSolverCreate(_comm->get(), &_ts);
+  DM da;
+  TSGetDM(_ts, &da);
+
+  DMCreateGlobalVector(da, &_petsc_sol);
+
+  // This requires because libMesh incorrectly treats the PETSc parallel vector as a ghost vector
```
Comment (lindsayad) [inline modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: "This requires" -> "This is required"

Have you talked with @roystgnr about this incorrect behavior?

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C [hunk 4]
```diff
@@ -48,6 +58,28 @@ ExternalPetscSolverApp::registerAll(Factory & f, ActionFactory & af, Syntax & /*
   /* register custom execute flags, action syntax, etc. here */
 }
 
+std::shared_ptr<Backup>
+ExternalPetscSolverApp::backup()
+{
+  auto backup = MooseApp::backup();
+
+  PetscVector<Number> petsc_sol(_petsc_sol, comm());
```
Comment (lindsayad) [inline modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: What's the reason for this copy construction? I'm sure you have one.

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C [hunk 5]
```diff
@@ -48,6 +58,28 @@ ExternalPetscSolverApp::registerAll(Factory & f, ActionFactory & af, Syntax & /*
...
+
+  dataStore(backup->_system_data, static_cast<NumericVector<Number> &>(petsc_sol), NULL);
```
Comment (lindsayad) [inline modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: NULL -> nullptr for C++

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: modules/external_petsc_solver/include/problem/ExternalPETScProblem.h (hunk 1 above)
Comment (fdkong) [reply modules/external_petsc_solver/include/problem/ExternalPETScProblem.h]: I want app to hold solution. Right, we could simply use a pointer. We could simply consider `Vec` as a class, and more readable for me.

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C (hunk 3 above)
Comment (fdkong) [reply modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: I know what was the problem.  But I did not see an easy to fix it.

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C (hunk 2 above)
Comment (fdkong) [reply modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: Copied from @loganharbour

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C (hunk 4 above)
Comment (fdkong) [reply modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: Just want to temporally add a wrapper to `_petsc_sol` (PETSc Vec), and we can use moose routines.  This is a not  copy constructor.  We construct a `PetscVector<Number>` using a `Vec` as input

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C (hunk 5 above)
Comment (fdkong) [reply modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: Will do

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C (hunk 3 above)
Comment (fdkong) [reply modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: And I do not want these change depend on another libmesh update.

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C (hunk 2 above)
Comment (loganharbour) [reply modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: ¯\_(ツ)_/¯

File: modules/external_petsc_solver/include/problem/ExternalPETScProblem.h (hunk 1 above)
Comment (lindsayad) [reply modules/external_petsc_solver/include/problem/ExternalPETScProblem.h]: So you want the `App` to hold the solution and not the `Problem`? And you're arguing that I shouldn't make use of my  knowledge that `Vec` is actually a pointer? That's a fair argument. Ok I support this change.

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C (hunk 3 above)
Comment (lindsayad) [reply modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: It's fair to not want to wait. But eventually if you think something is wrong at a lower level, then we should get it right at the lower level.

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C (hunk 4 above)
Comment (lindsayad) [reply modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: Ok duh, forgot that `_petsc_sol` was just a `Vec`. 👍

File: modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C (hunk 3 above)
Comment (fdkong) [reply modules/external_petsc_solver/src/base/ExternalPetscSolverApp.C]: I agreed. I added a PR here https://github.com/libMesh/libmesh/pull/2461. 

We do not need this line of code once libmesh is updated in MOOSE. 

But let us keep this line of code as is right now so that we can merge this change without waiting for another libmesh update.

**Review by lindsayad — APPROVED** [review]
